### PR TITLE
Refactored diffusion sampler

### PR DESCRIPTION
**Pull request by CharlelieLrt**
# PhysicsNeMo Pull Request

### Overview

This PR introduces a new modular framework for diffusion model sampling, designed to decouple the core components of the sampling process for maximum flexibility and reusability.

#### Decoupled Architecture

- **Model-agnostic**: The sampling function now works with any model implementing the `DiffusionDenoiser` protocol—shape-agnostic, dimension-agnostic, and supports arbitrary conditioning/guidance
- **Solver ↔ Noise Schedule decoupling**: Numerical solvers and noise schedules are now independent, allowing any combination
- **Flexible time-steps**: Time-step discretization can optionally be overridden independently of both solver and noise schedule

##### `Solver` Protocol (`physicsnemo.diffusion.samplers.solvers`)
Defines a minimal interface for numerical integration methods; 4 concrete solver implementations: Euler, Heun, and their respective EDM-stochastic versions.

##### `NoiseScheduler` Protocol (`physicsnemo.diffusion.noise_schedulers`)
Defines minimal interface for training (`add_noise`, `sample_time`) and sampling (`timesteps`, `init_latents`, `get_denoiser`). A base class (`LinearGaussianNoiseScheduler`) implements the protocol for the family of linear-Gaussian schedules; 6 concrete implementation of such noise schedulers: `LinearGaussianNoiseScheduler`, `EDMNoiseScheduler`, `VENoiseScheduler`, `VPNoiseScheduler`, `IDDPMNoiseScheduler`, `StudentTEDMNoiseScheduler`.

##### `DiffusionDenoiser` Protocol (`physicsnemo.diffusion.base`)
Minimal interface for denoisers.

##### New `sample` Function (`physicsnemo.diffusion.samplers.sample`)
Unified entry point for sampling with:
- Configurable solver (string key or custom instance)
- Configurable noise scheduler
- Optional custom time-steps
- Support for intermediate sample collection

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: physicsnemo/diffusion/noise_schedulers/noise_schedulers.py [hunk 1]
```diff
@@ -14,14 +14,1730 @@
...
+            ``num_steps``.
+        """
+        step_indices = torch.arange(num_steps, dtype=dtype, device=device)
+        smax_inv_rho = self.sigma_max ** (1 / self.rho)
+        smin_inv_rho = self.sigma_min ** (1 / self.rho)
+        frac = step_indices / (num_steps - 1)
+        interp = smax_inv_rho + frac * (smin_inv_rho - smax_inv_rho)
+        t_steps = interp**self.rho
+        zero = torch.zeros(1, dtype=dtype, device=device)
+        return torch.cat([t_steps, zero])
```
Comment (pzharrington) [inline physicsnemo/diffusion/noise_schedulers/noise_schedulers.py:1017]: So the idea is, if I want to use a different `sigma_min` and `sigma_max`, I either create a new `EDMNoiseScheduler` or I subclass this one and write a new `timesteps`/`sample_times` method for something more custom? I was thinking it might be convenient to pass `sigma_min` and `sigma_max` as kwargs to override the default defined `self.sigma_min`/`self.sigma_max`. Maybe that's opening up too much potential for bugs though

Comment (pzharrington) [reply physicsnemo/diffusion/noise_schedulers/noise_schedulers.py:1017]: Basically trying to think through the steps for something like StormScope sampling, which uses the `S_churn` setting from EDM sampling in addition to swapping from one pretrained denoiser to handle `(sigma_max, sigma_middle)` steps and then a different one for `(sigma_middle, sigma_min)` steps

File: physicsnemo/diffusion/samplers/samplers.py [hunk 2]
```diff
@@ -14,13 +14,337 @@
...
+    "edm_stochastic_heun": EDMStochasticHeunSolver,
+}
+
+
+def sample(
+    denoiser: DiffusionDenoiser,
+    xN: Float[Tensor, " B *dims"],
+    noise_scheduler: NoiseScheduler,
+    num_steps: int,
+    solver: str | Solver = "heun",
```
Comment (pzharrington) [inline physicsnemo/diffusion/samplers/samplers.py]: Update to  a `Literal` for only the approved solvers?

File: physicsnemo/diffusion/noise_schedulers/noise_schedulers.py (hunk 1 above)
Comment (CharlelieLrt) [reply physicsnemo/diffusion/noise_schedulers/noise_schedulers.py:1017]: Yeah to change any parameter you would need to either:
1. Re-instantiate a new noise scheduler: `scheduler = EDMNoiseScheduler(sigma_min=my_new_sigma_min, sigma_max=my_new_sigma_max)`
2. Simply update the attributes: `scheduler.sigma_max = my_new_sigma_max`. A noise scheduler is only a small container that stores a few numerical values. It does not have some kind of complex underlying behavior,  so kind of update is fine.

> I subclass this one and write a new timesteps/sample_times method for something more custom

You should **not** need to do that if it's only to specify a new `sigma_min` and `sigma_max`.


> Basically trying to think through the steps for something like StormScope sampling, which uses the S_churn setting from EDM sampling in addition to swapping from one pretrained denoiser to handle (sigma_max, sigma_middle) steps and then a different one for (sigma_middle, sigma_min) steps

I am not sure fully understand this. Maybe the best is you show me the math and we can chat offline about the best way to implement it.

File: physicsnemo/diffusion/samplers/samplers.py (hunk 2 above)
Comment (CharlelieLrt) [reply physicsnemo/diffusion/samplers/samplers.py]: Done in 1f59c4c2a8262b9d89e028ce3235a7d66461a765

File: physicsnemo/diffusion/base.py [hunk 3]
```diff
@@ -98,3 +98,100 @@ def __call__(
...
+
+    A denoiser is a callable that takes a noisy state ``x`` and a noise level
+    (or diffusion time) ``t``, and typically returns a denoising term. A
+    denoising could be the right hand side for
+    ODE/SDE-based sampling, directly the denoised latent state for discrete
+    Markov chain-based sampling, etc. This interface is generic and it does not
+    make any assumption about the nature of the denoising term. It is expected
+    to be used in conjunction with a compatible
+    :class:`~physicsnemo.diffusion.samplers.solvers.Solver` to perform the
+    actual sampling.
```
Comment (simonbyrne) [inline physicsnemo/diffusion/base.py]: Thanks, this is great.

**Review by NickGeneva — APPROVED** [review]
Approving to unblock, pending small usability improvements for the get_denoiser + adding (or renaming) a DiffusionOperator call to help clear up type confusion.
